### PR TITLE
Preprocessing label flipping

### PR DESCRIPTION
**Pull request by reluzita**

### CODE REVIEW

File: src/aequitas/flow/methods/preprocessing/label_flipping.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        
+        Parameters
+        ----------
+        flip_rate : float, optional
+            Maximum fraction of the training data to flip, by default 0.1
+        bagging_max_samples : float, optional
+            The number of samples to draw from X to train each base estimator of the 
+            bagging classifier (with replacement).
+        base_estimator : str, optional
+            The base estimator to fit on random subsets of the dataset. If None, the
```
Comment (sgpjesus) [inline src/aequitas/flow/methods/preprocessing/label_flipping.py:37]: If `None`, I believe this will be an error. This is a default! :stuck_out_tongue:

File: src/aequitas/flow/methods/preprocessing/label_flipping.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            classifiers predictions. By default "ensemble_margin".
+        """
+        self.logger = create_logger("methods.preprocessing.LabelFlipping")
+        self.logger.info("Instantiating a LabelFlipping preprocessing method.")
+
+        self.flip_rate = flip_rate
+        self.bagging_max_samples = bagging_max_samples
+
+        base_estimator = import_object(base_estimator)
+        args = {arg: value for arg, value in base_estimator_args.items() if arg in inspect.signature(base_estimator).parameters}
```
Comment (sgpjesus) [inline src/aequitas/flow/methods/preprocessing/label_flipping.py:60]: I think that we can log the ones we pass and the ones that are discarded here!

File: src/aequitas/flow/methods/preprocessing/label_flipping.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+class LabelFlipping(PreProcessing):
+    def __init__(
+            self,
+            flip_rate: float = 0.1,
+            bagging_max_samples: float = 0.5,
+            base_estimator: str = "sklearn.tree.DecisionTreeClassifier",
+            base_estimator_args: dict = {}, 
+            n_estimators: int = 10,
+            fair_ordering: bool = True,
+            ordering_method: str = "ensemble_margin",
```
Comment (sgpjesus) [inline src/aequitas/flow/methods/preprocessing/label_flipping.py:23]: One tip, for these types of variables with a fixed number of possible values, you can use literals:  https://peps.python.org/pep-0586/

I have not used it for everything tbh, but it might be useful!

File: src/aequitas/flow/methods/preprocessing/label_flipping.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+    def transform(self, X: pd.DataFrame, y: pd.Series, s: Optional[pd.Series]) -> Tuple[pd.DataFrame, pd.Series, Optional[pd.Series]]:
+        self.logger.info("Transforming data with LabelFlipping.")
+
+        if s is None and self.fair_ordering:
+            raise ValueError("Sensitive Attribute `s` not passed. Must be passed if `fair_ordering` is True.")
+        
+        X_num = pd.get_dummies(X)
+
+        bagging = BaggingClassifier(estimator=self.base_estimator,
```
Comment (sgpjesus) [inline src/aequitas/flow/methods/preprocessing/label_flipping.py:128]: Since we are fitting stuff here, I think we can do it in the `fit` method too. You have to save the `BaggingClassifier` object as an attribute of this class, but it will be fine, that way you can transform multiple times!

**Review by sgpjesus — COMMENTED** [review]
First round of comments!

File: src/aequitas/flow/methods/preprocessing/label_flipping.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+    def fit(self, X: pd.DataFrame, y: pd.Series, s: Optional[pd.Series]) -> None:
+        pass
+
+    def _score_instances(self, X, y, estimators):
+
+        if self.ordering_method == "ensemble_margin":
+            scores = pd.Series(dtype=float)
+            y_pred = np.array([clf.predict(X.values) for clf in estimators])
+            for i in X.index:
```
Comment (sgpjesus) [inline src/aequitas/flow/methods/preprocessing/label_flipping.py:79]: There are probably computationally faster ways of doing this loop. Maybe copilot can help you with it?

File: src/aequitas/flow/methods/preprocessing/label_flipping.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        
+        elif self.ordering_method == "residuals":
+            y_pred = np.array([abs(y - clf.predict_proba(X.values)[:,1]) for clf in estimators])
+            scores = pd.Series(y_pred.sum(axis=0) / self.n_estimators, index=X.index)
+
+        return scores
+    
+    def _calculate_prevalence_disparity(self, y: pd.Series, s: pd.Series):
+        prevalence_0 = y.loc[s == 0].value_counts()[1] / y.shape[0]
+        prevalence_1 = y.loc[s == 1].value_counts()[1] / y.shape[0]
```
Comment (sgpjesus) [inline src/aequitas/flow/methods/preprocessing/label_flipping.py:95]: We have to discuss ways of doing this step in a more generic way!

File: src/aequitas/flow/methods/preprocessing/label_flipping.py [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+class LabelFlipping(PreProcessing):
+    def __init__(
+            self,
+            flip_rate: float = 0.1,
+            bagging_max_samples: float = 0.5,
+            base_estimator: str = "sklearn.tree.DecisionTreeClassifier", 
+            n_estimators: int = 10,
+            fair_ordering: bool = True,
+            ordering_method: Literal["ensemble_margin", "residuals"] = "ensemble_margin",
+            seed: int = 42,
```
Comment (sgpjesus) [inline src/aequitas/flow/methods/preprocessing/label_flipping.py:23]: Unawareness parameter?

File: src/aequitas/flow/methods/preprocessing/label_flipping.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+from ...utils.imports import import_object
+
+import inspect
+import pandas as pd
+from typing import Optional, Tuple, Literal, Union, Callable
+import numpy as np
+from sklearn.ensemble import BaggingClassifier
+
+METHODS = ["ensemble_margin", "residuals"]
+
```
Comment (sgpjesus) [inline src/aequitas/flow/methods/preprocessing/label_flipping.py:13]: Some things, such as having two lines between a class and what comes before, is done automatically by a tool named `black`. Check for the extension `black Formatter` in VSCode, and change this setting: 
image
